roachtest: error: interface conversion: tree.Datum is tree.dNull, not *tree.DTime

roachtest.sqlsmith/setup=empty/setting=no-mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6698314?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6698314?buildTab=artifacts#/sqlsmith/setup=empty/setting=no-mutations) on master @ [4780ee15194e7d22b661ac92254b13cd2b71dcb1](https://github.com/cockroachdb/cockroach/commits/4780ee15194e7d22b661ac92254b13cd2b71dcb1):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
	sqlsmith.go:266,sqlsmith.go:327,test_runner.go:928: error: pq: internal error: interface conversion: tree.Datum is tree.dNull, not *tree.DTime
		stmt:
		WITH
			with_82539 (col_458866)
				AS (
					SELECT
						*
					FROM
						(VALUES (1001966053869705534:::INT8), ((-5769194913384792627):::INT8), ((-424029758473933114):::INT8))
							AS tab_266058 (col_458866)
				)
		SELECT
			tab_266073.col_458885 AS col_458886
		FROM
			(
				VALUES
					(7200918216798451660:::INT8),
					((3579286148060936003:::INT8::INT8 - (-3114191080537755598):::INT8::INT8)::INT8),
					((-3406484482343978236):::INT8),
					(NULL)
			)
				AS tab_266073 (col_458885)
		WHERE
			(((parse_time(e'D~<\x0bjN"@y':::STRING::STRING)::TIME::TIME - '50 years 646 days 17:22:33.314038':::INTERVAL::INTERVAL)::TIME <= (NULL::TIME + '-77 years -4 mons -205 days -21:21:14.334992':::INTERVAL::INTERVAL)::TIME) AND (postgis_dropbbox('01020000400200000080F75B01C6B59941386F7BE6AEDBE44154B099823B6200423A054B9FB934F3C17C2E3EA22757E84180A559A8951A9041':::GEOMETRY::GEOMETRY)::GEOMETRY::GEOMETRY IN ('010200000000000000':::GEOMETRY, NULL, st_rotate('01060000400200000001030000400100000004000000BA5D54EC2115F8418CFE85A1E66001C2E04CAB943D60C2C18075A191651EE841B0A4C5D8AED3E84130E9B499A49DFFC1A409D10A416EE1C1AC4AEF8650A2D7C12C8DE3F5FFEAE9C1BA5D54EC2115F8418CFE85A1E66001C2E04CAB943D60C2C1010300004001000000070000008077EF556D0EC9C160D16935B29A02C2FA636FAB3601F4C1503C763153A8C241CA237A5E7065FEC138C21A4FBF1DD6C180A7292340EBF141FB0ECB809277F3C109836785A74EF1C1B4D3EA07A7A9E24166C5993348C10142683AD1F36F8AFE41967B94462F1A02C2D84F572278F7D9416C2566667589EF41B2A4BDD74334FBC14048F81C67A2BC41FEE1FBF2DBD2E2C18077EF556D0EC9C160D16935B29A02C2FA636FAB3601F4C1':::GEOMETRY::GEOMETRY, 1.241244671529434:::FLOAT8::FLOAT8, '0103000040010000000B00000039D5F7CC4F0D00C24BE085637548F4C12EF07B86AC88F94152557F19EEABFAC1C6E684FFB62AF8C1E072CA2AD0DCE741E03E9A99787CEAC18242F6FF9276F4C1AE5E08896419F441DC45E2628B7DF44104C83462BF0101C278FD290DBBFDDA41902814B6BEDFFB41869AA1050D88E7C1A02BFF84738DFF4134F0D2C5E2A7ED410C783C00CB16F341BEA4AC17E1DDEEC120F97CDC791FC8413287922F2C3BF441E0610E29BA12C5C1F0EECC121737CD417CE31BDFA32B0242FC80E302762601C29C9B818F44E6D5C158294EBFE527F141F8D06898C6DE0142C2D3E738ADF7F8C15CC81268ABECF641006B0A19F6FAAB4139D5F7CC4F0D00C24BE085637548F4C12EF07B86AC88F941':::GEOMETRY::GEOMETRY)::GEOMETRY)))
		LIMIT
			68:::INT8;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-20077

--- a/pkg/sql/opt/norm/comp_funcs.go
+++ b/pkg/sql/opt/norm/comp_funcs.go
@@ -76,6 +76,9 @@ func (c *CustomFuncs) FoldBinaryCheckOverflow(
 	}
 
 	lDatum, rDatum := memo.ExtractConstDatum(left), memo.ExtractConstDatum(right)
+	if lDatum == tree.DNull || rDatum == tree.DNull {
+		return nil, false
+	}
 	result, err := eval.BinaryOp(c.f.evalCtx, o.EvalOp, lDatum, rDatum)
 	if err != nil {
 		return nil, false

--- a/pkg/sql/opt/norm/testdata/rules/comp
+++ b/pkg/sql/opt/norm/testdata/rules/comp
@@ -207,6 +207,17 @@ select
       ├── (s:4::TIME + '05:00:00') < '01:00:00' [outer=(4), stable]
       └── (s:4::TIME + '-05:00:00') < '23:00:00' [outer=(4), stable]
 
+# Regression test for #89024 - don't attempt to evaluate op for NULL values.
+norm expect-not=(NormalizeCmpPlusConst,NormalizeCmpMinusConst,NormalizeCmpConstMinus)
+SELECT 1
+WHERE (parse_time(e'D~<\x0bjN"@y')::TIME - '50 years'::INTERVAL)::TIME <= NULL::TIME
+----
+values
+ ├── columns: "?column?":1!null
+ ├── cardinality: [0 - 0]
+ ├── key: ()
+ └── fd: ()-->(1)
+
 # --------------------------------------------------
 # NormalizeCmpMinusConst
 # --------------------------------------------------

